destination/writer/writer.go

```diff
@@ -0,0 +1,251 @@
+	}
+
+	structuredData := make(sdk.StructuredData)
+	if err := json.Unmarshal(data.Bytes(), &structuredData); err != nil {
+		return nil, fmt.Errorf("unmarshal data into structured data: %w", err)
+	}
+
+	// convert keys to lower case
+	structuredDataLower := make(sdk.StructuredData)
+	for key, value := range structuredData {
+		if parsedValue, ok := value.(map[string]any); ok {
+			jsonValue, err := json.Marshal(parsedValue)
+			if err != nil {
+				return nil, fmt.Errorf("marshal map into json: %w", err)
+			}
+
+			structuredDataLower[strings.ToLower(key)] = string(jsonValue)
+
+			continue
+		}
+
+		structuredDataLower[strings.ToLower(key)] = value
+	}
+
+	return structuredDataLower, nil
+}
+
+// extractColumnsAndValues turns the payload into slices of
+// columns and values for upserting into Firebolt.
+func (w *Writer) extractColumnsAndValues(payload sdk.StructuredData) ([]string, []any) {
+	var (
+		columns []string
+		values  []any
+	)
+
+	for key, value := range payload {
+		columns = append(columns, key)
+		values = append(values, value)
+	}
+
+	return columns, values
+}
+
+// convertPayload converts a sdk.StructureData values to a proper database types.
+func (w *Writer) convertPayload(data sdk.StructuredData) (sdk.StructuredData, error) {
+	result := make(sdk.StructuredData, len(data))
+
+	for key, value := range data {
+		if value == nil {
+			result[key] = value
+
+			continue
+		}
+
+		switch reflect.TypeOf(value).Kind() {
+		case reflect.Map, reflect.Slice:
+			bs, err := json.Marshal(value)
+			if err != nil {
+				return nil, fmt.Errorf("marshal: %w", err)
+			}
+
+			result[key] = string(bs)
+
+			continue
+		}
+
+		switch w.columnTypes[strings.ToLower(key)] {
+		case typeDate:
+			v, ok := value.(time.Time)
+			if ok {
+				result[key] = v.Format("2006-01-02")
+
+				continue
+			}
+
+			valueStr, ok := value.(string)
+			if ok {
+				timeValue, err := parseToTime(valueStr)
+				if err != nil {
+					return nil, fmt.Errorf("convert value to time.Time: %w", err)
+				}
+
+				result[key] = timeValue.Format("2006-01-02")
+
+				continue
+			}
+
+			return nil, ErrInvalidTypeForDateColumn
+		case typeTimestamp:
+			v, ok := value.(time.Time)
+			if ok {
+				// firebolt date type support this format
+				// https://docs.firebolt.io/general-reference/data-types.html#date-and-time
+				result[key] = v.Format("2006-01-02 15:04:05")
+
+				continue
+			}
+
+			valueStr, ok := value.(string)
+			if ok {
+				timeValue, err := parseToTime(valueStr)
+				if err != nil {
+					return nil, fmt.Errorf("convert value to time.Time: %w", err)
+				}
+
+				// firebolt timestamp type support this format
+				// https://docs.firebolt.io/general-reference/data-types.html#timestamp
+				result[key] = timeValue.Format("2006-01-02 15:04:05")
+
+				continue
+			}
+
+			return nil, ErrInvalidTypeForTimestampColumn
+		default:
+			result[key] = value
+		}
+	}
+
+	return result, nil
+}
+
+func parseToTime(val string) (time.Time, error) {
```

Comment:
Maybe make this a receiver method (if it's only used by the writer)?

Reply:
yeah, good idea. Done